flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/CatalogSchemaTable.java

```diff
@@ -153,6 +166,27 @@ private static RelDataType getRowType(RelDataTypeFactory typeFactory,
 				}
 			}
 		}
+
+		// The following block is a workaround to support tables defined by TableEnvironment.connect() and
+		// the actual table sources implement DefinedProctimeAttribute/DefinedRowtimeAttributes.
+		// It should be removed after we remove DefinedProctimeAttribute/DefinedRowtimeAttributes.
+		Optional<TableSource> sourceOpt = findAndCreateTableSource(new TableConfig().getConfiguration());
+		if (tableSchema.getTableColumns().stream().noneMatch(TableColumn::isGenerated)
+			&& tableSchema.getWatermarkSpecs().isEmpty()
```

Comment:
Add `isStreamingMode` into this condition, and `findAndCreateTableSource`  when the condition is satisfied.